api/PayrollProcessor.Data.Persistence/Features/Employees/EmployeesQueryHandler.cs

```diff
@@ -0,0 +1,69 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Ardalis.GuardClauses;
+using LanguageExt;
+using Microsoft.Azure.Cosmos;
+using Microsoft.Azure.Cosmos.Linq;
+using PayrollProcessor.Core.Domain.Features.Employees;
+using PayrollProcessor.Core.Domain.Intrastructure.Operations.Queries;
+
+using static LanguageExt.Prelude;
+
+namespace PayrollProcessor.Data.Persistence.Features.Employees
+{
+    public class EmployeesQueryHandler : IQueryHandler<EmployeesQuery, IEnumerable<Employee>>
+    {
+        private readonly CosmosClient client;
+
+        public EmployeesQueryHandler(CosmosClient client)
+        {
+            Guard.Against.Null(client, nameof(client));
+
+            this.client = client;
+        }
+
+        public TryOptionAsync<IEnumerable<Employee>> Execute(EmployeesQuery query, CancellationToken token = default)
+        {
+            var (count, firstName, lastName) = query;
+
+            var dataQuery = client
+                .EmployeesQueryable<EmployeeRecord>()
+                .Where(e => e.Type == nameof(EmployeeRecord));
+
+            if (!string.IsNullOrWhiteSpace(firstName))
+            {
+                dataQuery = dataQuery.Where(e => e.FirstNameLower.Contains(firstName.ToLowerInvariant()));
+            }
+
+            if (!string.IsNullOrWhiteSpace(lastName))
+            {
+                dataQuery = dataQuery.Where(e => e.LastNameLower.Contains(lastName.ToLowerInvariant()));
+            }
+
+            if (count > 0)
+            {
+                dataQuery = dataQuery.Take(count);
```

Comment:
I found out through testing that the order of LINQ expressions matters with Cosmos.

The `.Take()` call has to be the last, otherwise the `CosmosClient` interprets the LINQ expressions that follow as being part of a sub-query, which isn't allowed to be combined with `.Take()`.

Reply:
Good to know, i'll have to play around with the queries to gain more understanding of them.